File: vane-core/src/main/java/org/oddlama/vane/core/command/check/ErrorCheckResult.java
package org.oddlama.vane.core.command.check;

import java.util.Objects;
import org.bukkit.command.CommandSender;
import org.oddlama.vane.core.command.Command;

public class ErrorCheckResult implements CheckResult {

    private int depth;
    private String message;
    private String argChain = "";

    public ErrorCheckResult(int depth, String message) {
        this.depth = depth;
        this.message = message;
    }

    @Override
    public int depth() {
        return depth;
    }

    @Override
    public boolean good() {
        return false;
    }

    @Override
    public boolean apply(Command<?> command, CommandSender sender) {
        return apply(command, sender, "");
    }

    public boolean apply(Command<?> command, CommandSender sender, String indent) {
        var str = indent;
        if (Objects.equals(indent, "")) {
            str += "§cerror: ";
        }
        str += "§6";
        str += argChain;
        str += message;
        sender.sendMessage(str);
        return false;
    }

    @Override
    public CheckResult prepend(String argumentType, Object parsedArg, boolean include) {
        // Save parsed arguments in an argument chain, and propagate error
        argChain = "§3" + argumentType + "§6 → " + argChain;
        return this;
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/command/check/ErrorCheckResult.kt
package org.oddlama.vane.core.command.check

import org.bukkit.command.CommandSender
import org.oddlama.vane.core.command.Command

/**
 * Failed parse/match result carrying an error message.
 *
 * @param depth parse depth where the failure occurred.
 * @param message localized or formatted error message.
 */
open class ErrorCheckResult(private val depth: Int, private val message: String) : CheckResult {
    /**
     * Rendered chain of argument types matched before this error.
     */
    private var argChain = ""

    /** Returns the parse depth for this result. */
    override fun depth(): Int = depth

    /** Error results are always unsuccessful. */
    override fun good(): Boolean = false

    /** Applies this error using root indentation. */
    override fun apply(command: Command<*>?, sender: CommandSender?): Boolean =
        apply(command, sender, "")

    /** Sends this error message to the sender. */
    open fun apply(command: Command<*>?, sender: CommandSender?, indent: String): Boolean {
        val prefix = indent.ifEmpty { "${indent}§cerror: " }
        sender?.sendMessage("${prefix}§6${argChain}${message}")
        return false
    }

    /** Prepends argument context to the displayed error chain. */
    override fun prepend(argumentType: String, parsedArg: Any?, include: Boolean): CheckResult {
        argChain = "§3${argumentType}§6 → $argChain"
        return this
    }
}
